### PR TITLE
Type annotations, mypy config, pydoclint setup, and other changes

### PR DESCRIPTION
**Pull request by VeckoTheGecko**
## Overview

This PR introduces typing to the Parcels codebase, providing:

- useful information to users regarding the types of parameters
- useful information to developers regarding the types of variables
- a mechanism to bring types into active consideration, where complicated typing can be a "code smell" indicating a deeper problem with the code

The coverage for this isn't 100% yet (as doing so is quite difficult with the current codebase), but as we refactor code it will get easier to increase our coverage.

## Changes
Typing:
- Add mypy to check types (also with a CI workflow)
- Add mypy config, with list of files to check for types (this list will be included). Note that this list is only the source files (i.e., entrypoint for mypy). Types with imported `parcels` functions/methods are also checked
- Add typing to various files
- Introduce a `_typing.py` file to define types used throughout Parcels

Other:
- add `pydoclint` tooling to pre-commit to help capture docstring errors not picked up by Ruff
- add `_compat.py` file to help with compatibility between installations when using mpi
- Implement some suggestions from [repo-review tool](https://learn.scientific-python.org/development/guides/repo-review/?repo=oceanparcels%2Fparcels&branch=v%2Ftype-annotations)
  - Notably:
    - concurrency with GHA (i.e., repeated pushes to a branch will cancel already running workflows on prior commits)
    - dependabot to automatically open PRs to update GHA workflow versions monthly
- Fix unit-test badge in README and add Ruff badge
- Clean up some code involving casting to arrays

### CODE REVIEW

File: parcels/_typing.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#     "nearest",
+#     "freeslip",
+#     "partialslip",
+#     "bgrid_velocity",
+#     "bgrid_w_velocity",
+#     "cgrid_velocity",
+#     "linear_invdist_land_tracer",
+#     "nearest",
+#     "cgrid_tracer",
+# ] # mostly corresponds with `interp_method` # TODO: This should be narrowed. Unlikely applies to every context
```
Comment (VeckoTheGecko) [inline parcels/_typing.py:30]: @erikvansebille do you have insight into the contexts in which different interpolation methods are used? Is there additional nuance here?

Comment (erikvansebille) [reply parcels/_typing.py:30]: What do you mean with additional nuance? See https://docs.oceanparcels.org/en/latest/examples/tutorial_interpolation.html for an overview of (almost?) all of these options, and how they differ.

File: parcels/fieldset.py [hunk 2]
```diff
@@ -136,7 +134,7 @@ def from_data(
             lat = dims["lat"]
             depth = np.zeros(1, dtype=np.float32) if "depth" not in dims else dims["depth"]
             time = np.zeros(1, dtype=np.float64) if "time" not in dims else dims["time"]
-            time = np.array(time) if not isinstance(time, np.ndarray) else time
+            time = np.array(time)
```
Comment (erikvansebille) [inline parcels/fieldset.py:137]: So I assume this is a valid change because calling `np.array()` on an `np.array` doesn't change anything to the array?

File: parcels/grid.py [hunk 3]
```diff
@@ -3,7 +3,9 @@
 from enum import IntEnum
 
 import numpy as np
+import numpy.typing as npt
```
Comment (erikvansebille) [inline parcels/grid.py:6]: Why don't we use `npt` in the other files too? This is the first I've now come across during this review?

File: parcels/kernel.py [hunk 4]
```diff
@@ -139,6 +136,9 @@ def remove_deleted(self, pset):
             self.fieldset.particlefile.write(pset, None, indices=indices)
         pset.remove_indices(indices)
 
+    @abc.abstractmethod
+    def get_kernel_compile_files(self): ...
+
```
Comment (erikvansebille) [inline parcels/kernel.py:141]: Why is this here? What does this do?

File: parcels/tools/converters.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 import cftime
 import numpy as np
 import xarray as xr
+from numpy.typing import ArrayLike, NDArray
```
Comment (erikvansebille) [inline parcels/tools/converters.py:9]: Why not use `import numpy.typing as npt` like in grid.py? Better to stay consistent throughout Parcels?

File: parcels/_typing.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+
+KernelFunction = Callable[..., None]
+
+
+def ensure_is_literal_value(value: Any, literal: Any) -> None:
+    """Ensures that a value is a valid option for the provided Literal type annotation."""
+    valid_options = get_args(literal)
+    if value not in valid_options:
+        raise ValueError(f"{value!r} is not a valid option. Valid options are {valid_options}")
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/_typing.py:50]: Will remove and put into future PR

File: parcels/_typing.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+PathLike = str | os.PathLike
+Mesh = Literal["spherical", "flat"]  # mostly corresponds with `mesh`
+VectorType = Literal["3D", "2D"] | None  # mostly corresponds with `vector_type`
+ChunkMode = Literal["auto", "specific", "failsafe"]  # mostly corresponds with `chunk_mode`
+GridIndexingType = Literal["pop", "mom5", "mitgcm", "nemo"]  # mostly corresponds with `grid_indexing_type`
+UpdateStatus = Literal["not_updated", "first_updated", "updated"]  # mostly corresponds with `update_status`
+TimePeriodic = float | datetime.timedelta | Literal[False]  # mostly corresponds with `update_status`
+NetcdfEngine = Literal[
+    "netcdf4", "xarray"
+]  # TODO: It seems that "scipy" is also an option (according to a docstring) but can't find mention in code. Investigate.
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/_typing.py:40]: add scipy and mention in docstrings

File: parcels/_typing.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+#     "partialslip",
+#     "bgrid_velocity",
+#     "bgrid_w_velocity",
+#     "cgrid_velocity",
+#     "linear_invdist_land_tracer",
+#     "nearest",
+#     "cgrid_tracer",
+# ] # mostly corresponds with `interp_method` # TODO: This should be narrowed. Unlikely applies to every context
+PathLike = str | os.PathLike
+Mesh = Literal["spherical", "flat"]  # mostly corresponds with `mesh`
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/_typing.py:32]: remove 'mostly'

File: parcels/_typing.py (hunk 1 above)
Comment (VeckoTheGecko) [reply parcels/_typing.py:30]: activate code and add type annotations

File: parcels/fieldset.py (hunk 2 above)
Comment (VeckoTheGecko) [reply parcels/fieldset.py:137]: yes :)

File: parcels/grid.py (hunk 3 above)
Comment (VeckoTheGecko) [reply parcels/grid.py:6]: yes, can do

File: parcels/field.py [hunk 9]
```diff
@@ -854,7 +870,9 @@ def search_indices_vertical_z(self, z):
         zeta = (z - grid.depth[zi]) / (grid.depth[zi + 1] - grid.depth[zi])
         return (zi, zeta)
 
-    def search_indices_vertical_s(self, x, y, z, xi, yi, xsi, eta, ti, time):
+    def search_indices_vertical_s(
+        self, x: float, y: float, z: float, xi: int, yi: int, xsi: float, eta: float, ti: int, time
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:874]: will add :float

File: parcels/field.py [hunk 10]
```diff
@@ -33,6 +35,11 @@
 )
 from .grid import CGrid, Grid, GridType
 
+if TYPE_CHECKING:
+    from ctypes import _Pointer as PointerType
+
+    from parcels.fieldset import FieldSet
+
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:42]: Missed this in our meeting. The `TYPE_CHECKING` boolean is set by `mypy` (and also editors like vscode) when it goes through to evaluate types. This allows you to expose things to mypy and type checkers which has some benefits:

- Avoiding circular imports (which is why its done here, `parcels.fieldset` imports `Field`, so if we have a `from parcels.fieldset import FieldSet` here we run into a circular import)
- Reducing runtime overhead slightly by avoiding an import if its only used for type hinting
- to set arbitrary custom behaviour when typechecking

Note that when this is done you have to use the string `"FieldSet"` instead of `FieldSet` as a type hint in order to avoid errors at runtime (but the type checking still works out)

File: parcels/field.py [hunk 11]
```diff
@@ -637,7 +646,14 @@ def from_netcdf(
 
     @classmethod
     def from_xarray(
-        cls, da, name, dimensions, mesh="spherical", allow_time_extrapolation=None, time_periodic=False, **kwargs
+        cls,
+        da: xr.DataArray,
+        name: str,
+        dimensions,
+        mesh: Mesh = "spherical",
+        allow_time_extrapolation: bool | None = None,
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:654]: It's because it is set as the default option in the signature, and somewhere in the code must deal with the None case. I agree that this can be refactored in a future PR

[the remark this answers is not in the document]
Comment (erikvansebille) [reply parcels/field.py:654]: Hmm, would now not be a good opportunity to change the default to `False` throughout. 

I dug a little through the code, and I think the default is None because of this logic. 
https://github.com/OceanParcels/Parcels/blob/e146ff344f734fd44061512bfbcb84d7c43967af/parcels/field.py#L211-L214
But there must be another way to implement this?

File: parcels/field.py [hunk 12]
```diff
@@ -283,10 +291,10 @@ def __init__(
         # (data_full_zdim = grid.zdim if no indices are used, for A- and C-grids and for some B-grids). It is used for the B-grid,
         # since some datasets do not provide the deeper level of data (which is ignored by the interpolation).
         self.data_full_zdim = kwargs.pop("data_full_zdim", None)
-        self.data_chunks = []  # the data buffer of the FileBuffer raw loaded data - shall be a list of C-contiguous arrays
-        self.c_data_chunks = []  # C-pointers to the data_chunks array
-        self.nchunks = []
-        self.chunk_set = False
+        self.data_chunks = []  # type: ignore # the data buffer of the FileBuffer raw loaded data - shall be a list of C-contiguous arrays
```
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:294]: Yes, you read my mind. Another PR I have in mind is to privatise a lot of the attributes and methods we have. I have an idea to do so in a way that is minimally disruptive to users (by raising deprecation warnings) and also doesn't have maintenance burden.

File: parcels/field.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:654]: In that case, I think it makes sense. Where its treated as a static field if `allow_time_extrapolation` is None and the time dims are one. We could instead make it explicit that they have to set it to `True` for static fields?

```python
if not allow_time_extrapolation and len(self.grid.time) == 1: 
     raise ValueError("Explicitly set `allow_time_extrapolation` to True when working with static fields.")

self.allow_time_extrapolation = allow_time_extrapolation
```

[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:654]: I'm inclined to make this a separate issue, along with an issue about a Parcels deprecation policy

[the remark this answers is not in the document]
Comment (erikvansebille) [reply parcels/field.py:654]: OK, now I understand. Your proposed change would indeed have a major impact; not least on many, many unit tests with time-stationary fields. So I agree then it's not worth the change just for the slighter cleaner type-annotations

[the remark this answers is not in the document]
Comment (VeckoTheGecko) [reply parcels/field.py:654]: Cool, lets keep the current behaviour then

File: parcels/fieldfilebuffer.py [hunk 13]
```diff
@@ -9,14 +9,22 @@
 from dask import utils as da_utils
 from netCDF4 import Dataset as ncDataset
 
+from parcels._typing import InterpMethodOption
```
Comment (erikvansebille) [inline parcels/fieldfilebuffer.py:12]: Why do we call this `InterpMethodOption` and not `InterpMethod`? The other types also don't end in `...Option`

Comment (VeckoTheGecko) [reply parcels/fieldfilebuffer.py:12]: This is highlighted in the `_typing.py` file. `InterpMethod` also includes typing for dictionary mappings

**Comment by VeckoTheGecko** [comment]
Fixes #1653 . Refactoring to reduce code duplication in `fieldset.py` will be done in a future PR
